### ISSUE TITLE
What is the best practice way to handle observable @Input()

### ISSUE THREAD

**Issue by Laurens-makel** [rx-angular/rx-angular#1658]
# Description

The documentation provides the following statement in the tutorial part:
https://www.rx-angular.io/docs/state/tutorials/passing-observables-directly
![image](https://github.com/rx-angular/rx-angular/assets/37303445/3949b687-7849-471f-898b-fc5e0714264c)

This works but I've noticed there is a little problem regarding this. Whenever a new observable is passed by the Input(), the previously connected observable is never unsubscribed. This could cause unwanted results when any of the previous observables emits a new value after the latest observable did.

For example, a parent component has an array of objects and an active index number in it's state. These objects contain observable properties, for example `users$`. 

```typescript
export interface Item {
  users$: Observable<{name: string}>
}
export interface State {
  list: Item[]
  active: number
}
```

Parent composes an observable `active$`, which takes the active index number and the array to compose this value. Basically `active$.users$` is passed to the component referenced exposing the `@Input()`. 

```typescript
  active$ = this.state.select().pipe(
    selectSlice(['list', 'active']),
    map( ([list, active]) => list[active])
  )
``` 

```html
<child-component *rxLet="active$; let active" [users]="active.users$"></child-component>
```
The user decides to starts an async action which is supposed to update the current `users$` list and quickly decides to display another object, which changes the active number index.

The parent components `active$` is updated since the active index changed, a new observable is passed to the `Input()` becuase `active$.users$` refers to a new observable, the new value is rendered and after this the async action completes, making the previous observable emit again and a value is rendered which does not belong to the current `active$` because the previous is never unsubscribed.

One could work around this problem by using a `Subject` and `takeUntil()`, but this is stupid boilerplate we are trying to avoid:
```typescript
  @Input()
  set users(users$) {
    this.nextUsers.next(true);
    this.connect('users', users$.pipe(takeUntil(this.nextUsers)));
  }
  nextUsers = new Subject<true>();
```

or  something like:
```typescript
  @Input()
  set users(users$) {
    this.nextUsers.next(users);
  }
  nextUsers = new Subject<Observable<{name: string}>>();

  constructor() {
    this.state.connect('users', this.nextUsers.pipe( switchMap(users$=> users$) )
  }
```

This could be simplified with the use of rxActions, which basically does the same.

```typescript
export interface Actions {
  setUsers: Observable<{name: string}>
}
```
```typescript
  @Input()
  set users(users$) {
    this.actions.setUsers(users$);
  }

  constructor() {
    this.state.connect('users', this.actions.setUsers$.pipe( switchMap(users$=> users$) )
  }
```
or
```typescript
  @Input()
  set users(users$) {
    this.actions.setUsers(users$.pipe( switchMap(users$=> users$) );
  }

  constructor() {
    this.state.connect('users', this.actions.setUsers$)
  }
```

Another solution would be to not 'override' the `@Input()` reference by creating a single observable in the parent component by doing something like this:

```typescript
  activeUsers$ = this.active$.pipe(
    switchMap(item => item.users$)
  )
```
```html
<child-component [users]="activeUsers$"></child-component>
```

Although this approach is still vulnerable if the reference to `activeUsers$` is changed for some reason. 

I dont consider this as a 'bug' but I have difficulty trying to understand on how to properly deal with such a situation without too much boilerplate.

Could the documentation be updated with the above scenario in mind? Minimal a disclaimer that warns the user for this behavioru and Ideally with a recommend practice on how to deal with such a situation.

How do you possibly think about introducing a new method or override for the existting `connect` method to support discarding previous observables without any of above tricks?

```

# Steps to Reproduce the Issue

<!--
Explain detailed information about the steps how to reproduce this specific bug

In best case provide a [stackblitz](stackblitz.com) example

-->

# Environment

<!--
Include versions like the following as well as any additional context that's relevant to the issue:

- Version of the app or even better git commit id (get via `git log --oneline | head -n 1`): e.g. 123a456
- Browser version (get via Help»About): e.g. FF 77.0.1
- IDE and version (if relevant to the issue)
- Lib versions installed: (get all versions by running `yarn nx version`)
e.g.
```
Angular CLI: 10.0.5
Node: 12.16.2
OS: win32 x64

Angular: 10.0.7
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.1000.5
@angular-devkit/build-angular     0.1000.5
@angular-devkit/build-optimizer   0.1000.5
@angular-devkit/build-webpack     0.1000.5
@angular-devkit/core              10.0.5
@angular-devkit/schematics        10.0.5
@angular/cdk                      10.1.2
@angular/cli                      10.0.5
@angular/material                 10.1.2
@ngtools/webpack                  10.0.5
@schematics/angular               10.0.5
@schematics/update                0.1000.5
ng-packagr                        10.0.3
rxjs                              6.6.2
typescript                        3.9.7
webpack                           4.43.0
```
-->

# Related to Other Issues

<!--
If applicable, otherwise remove a section
-->

# Tasks to Resolve This

<!--
If a breakdown into steps required to fix the issue is already possible. Otherwise, remove this section.
-->

# Notes

<!--
Anything that didn't fit into the above categories. Remove this section if it isn't needed.
-->

**Comment by hoebbelsB** [comment]
thanks @Laurens-makel for the valuable input. You are right, passing an Observable as Input does not come without risks. As a best practice I would only pass Observables as Inputs, if the reference is not changing. And as you are working with observables, this should always be possible.
As you are anyway working with streams, you can pass new values to the stream instead of passing an entirely new stream of values.

Given your example:

> Parent composes an observable active$, which takes the active index number and the array to compose this value. Basically active$.users$ is passed to the component referenced exposing the @Input().

```ts
 active$ = this.state.select().pipe(
    selectSlice(['list', 'active']),
    map( ([list, active]) => list[active])
  )
```

```html
<child-component *rxLet="active$; let active" [users]="active.users$"></child-component>
```

I would refactor it to something like:

```ts
activeUsers$ = this.state.select().pipe(
    selectSlice(['list', 'active']),
    switchMap( ([list, active]) => list[active])
  )
```

```html
<child-component [users]="activeUsers$"></child-component>
```

In any case, I would try to keep one stream of values instead of changing the reference. 

Anyway, we need to make sure this is properly stated in the docs. Do you might want to contribute at this point? :)

**Comment by Laurens-makel** [comment]
@hoebbelsB 
Thanks for your reply which definitely clarifies my expectations. I've created a PR to clarify this behaviour in the docs.

What about introducing a fail safe for this?

I've developed an annotation to create an observable of observable `@Input()` fields, under the hood uses the `Subject` and `switchMap()` approach to have 'current' observable of truth. The usage of `switchMap` in this annotation causes previous passed observables to be unsubscribed in the `switchMap` operator.

Which allows changing the reference of the obserable and gives the following syntax.
```typescript
@Input() users: Observable<User[]>;
@Observe("users") usersInput: Observable<User[]>

constructor(){
  this.state.connect('users', this.usersInput);
}

```

Or would this go against architectural choices?
